### PR TITLE
Check manually-assigned addresses with IPAM to see if they clash

### LINKED ISSUE #687: [ipam] refuse manual allocation in ipam range
If a user inadvertently assigns an IP in the range owned by IPAM (as described in #681), weave could display an informative message.

### PR DESCRIPTION
**Pull request by bboreham**
Fixes #687 and #598

One quiet change in this PR:
 * avoid printing error messages from 'claim' twice in the logs

### CODE REVIEW

**Comment by rade** [comment]
> I could add some more lines to remove this if desired.

yes, please.

**Comment by bboreham** [comment]
Done

File: ipam/allocator.go [hunk 1]
```diff
@@ -215,9 +215,9 @@ func (alloc *Allocator) Lookup(ident string, r address.Range) (address.Address,
 }
 
 // Claim an address that we think we should own (Sync)
-func (alloc *Allocator) Claim(ident string, addr address.Address) error {
+func (alloc *Allocator) Claim(ident string, addr address.Address, launch bool) error {
 	resultChan := make(chan error)
-	op := &claim{resultChan: resultChan, ident: ident, addr: addr}
+	op := &claim{resultChan: resultChan, ident: ident, addr: addr, launch: launch}
```
Comment (squaremo) [inline ipam/allocator.go]: I do not understand what `launch` mean in these changes

Comment (bboreham) [inline ipam/allocator.go]: I was trying to convey that we want it to behave differently when called at launch of weave compared to later.

Comment (squaremo) [inline ipam/allocator.go]: I think it would be better to name it for how it behaves differently

**Comment by squaremo** [comment]
I think these are two different operations, rather than a variation on one operation (and perhaps that is why I was confused by "launch").

The first is to tell weave about an IP that had been claimed previously. The second is to ask for a particular IP that has not been handed out previously. It's actually the latter that deserves the name "Claim"; the former is more like, well, "Reclaim" or "Recover".

How this is reflected in the HTTP API, I'm not sure -- PUT to a URL seems a good match for both. So a query arg might be appropriate for discriminating.

**Comment by bboreham** [comment]
I renamed 'launch' to 'reclaim'; rebased and squashed.
Also changed the behaviour to output a message and carry on when IPAM is waiting for quorum, rather than hang the caller.  This is more in line with what #687 asks for.

File: ipam/claim.go [hunk 2]
```diff
@@ -12,6 +12,7 @@ type claim struct {
 	resultChan chan<- error
 	ident      string
 	addr       address.Address
+	reclaim    bool
 }
 
 func (c *claim) sendResult(result error) {
```
Comment (squaremo) [inline ipam/claim.go]: As a drive-by comment: I find it somewhat counter-intuitive that `claim.sendResult(nil)` indicates a successful operation.

File: ipam/claim.go [hunk 3]
```diff
@@ -40,15 +44,23 @@ func (c *claim) Try(alloc *Allocator) bool {
...
+		}
 		return false
 	default:
 		alloc.debugf("requesting address %s from other peer %s", c.addr, owner)
 		err := alloc.sendSpaceRequest(owner, address.NewRange(c.addr, 1))
-		if err != nil { // can't speak to owner right now; figure it out later
-			alloc.infof("Claim %s for %s: %s; will try later.", c.addr, c.ident, err)
-			c.sendResult(nil)
+		if err != nil { // can't speak to owner right now
+			if c.reclaim { // don't hold up launch; figure it out later
```
Comment (squaremo) [inline ipam/claim.go]: > // don't hold up launch; figure it out later

This comment doesn't make sense -- what does `reclaim` have to do with launch?

Comment (bboreham) [inline ipam/claim.go]: this code is only called from `weave launch`

Comment (squaremo) [inline ipam/claim.go]: that's a coincidence

Comment (squaremo) [inline ipam/claim.go]: Or think of it this way: if reclaim is a general-purpose operation that happens to be called at launch, this comment doesn't belong here. If reclaim is an operation intended _only_ for use during launch, why is it not better adapted to its purpose, e.g., a bulk operation (in other words, why does it look general purpose ..)

Comment (bboreham) [inline ipam/claim.go]: that's an accident

Comment (squaremo) [inline ipam/claim.go]: > that's an accident

which bit?

Comment (bboreham) [inline ipam/claim.go]: that it looks general-purpose.  Certainly the behaviour of apparently agreeing with a claim when we can't answer for sure yet (introduced in #1104) is purely directed at launch-time.

Comment (squaremo) [inline ipam/claim.go]: OK. As it is, it's a non-sequitur: "if reclaiming, don't hold up launch". Why should you have to know how it's used in the script to be able to make sense of that?

As much as I dislike the idea of holding things up for a comment, I think this is going to cause befuddlement for the next person to look at the bit of code. What I'd rather see is an explanation of why there is a different behaviour for an apparently very similar operation -- or for it to obviously be a different operation.

**Comment by bboreham** [comment]
I decided `noErrorOnUnknown` described the behaviour best.

File: ipam/claim.go [hunk 4]
```diff
@@ -40,15 +43,23 @@ func (c *claim) Try(alloc *Allocator) bool {
 		// success
 	case router.UnknownPeerName:
 		// If our ring doesn't know, it must be empty.
-		alloc.infof("Claim %s for %s: address allocator still initializing; will try later.", c.addr, c.ident)
-		c.sendResult(nil) // don't make the caller wait
+		if c.noErrorOnUnknown {
+			alloc.infof("Claim %s for %s: address allocator still initializing; will try later.", c.addr, c.ident)
+			c.sendResult(nil) // don't make the caller wait
+		} else {
+			c.sendResult(fmt.Errorf("%s is in the range %s, but the allocator is not initialized yet", c.addr, alloc.universe.AsCIDRString()))
```
Comment (squaremo) [inline ipam/claim.go]: The suggestion of the previous comment is that this will block the caller. Is that the case?

Comment (squaremo) [inline ipam/claim.go]: Whatever, I'll fix it up in post

Comment (squaremo) [inline ipam/claim.go]: > Whatever, I'll fix it up in post

#1219

**Comment by rade** [comment]
I struggle to see how this solves #598. Please explain.

**Comment by bboreham** [comment]
Line 866 in the `weave` script informs IPAM of a manually-selected address.  If this can be granted, it is, and hence IPAM will not then allocate that address to something else, which I thought was the main issue of #598.

**Comment by rade** [comment]
ok

**Comment by bboreham** [comment]
I should perhaps have updated the docs, where it says 

> make sure they don’t conflict with any IP ranges in use on the hosts (including those delegated to weave’s automatic IP address allocator)

**Comment by rade** [comment]
Surely http://docs.weave.works/weave/master/head/ipam.html#manual should be updated.
